aptos-move/move-examples/drand/sources/drand.move

```diff
@@ -0,0 +1,136 @@
+
+    /// The `bls-unchained-on-g1` drand beacon produces an output every 3 seconds.
+    /// (Or goes into catchup mode, if nodes fall behind.)
+    const PERIOD_SECS : u64 = 3;
+
+    /// The UNIX time (in seconds) at which the beacon started operating (this is the time of round #1)
+    const GENESIS_TIMESTAMP : u64 = 1677685200;
+
+    /// The drand beacon's PK, against which any beacon output for a round `i` can be verified.
+    const DRAND_PUBKEY : vector<u8> = x"a0b862a7527fee3a731bcb59280ab6abd62d5c0b6ea03dc4ddf6612fdfc9d01f01c31542541771903475eb1ec6615f8d0df0b8b6dce385811d6dcf8cbefb8759e5e616a3dfd054c928940766d9a5b9db91e3b697e5d70a975181e007f87fca5e";
+
+    /// The domain-separation tag (DST) used in drand's BLS signatures
+    const DRAND_DST: vector<u8> = b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_";
+
+    /// Error code for when anyone submits an incorrect randomness in our APIs (e.g., wrong-size).
+    const E_INCORRECT_RANDOMNESS: u64 = 1;
+
+    /// A randomness object, created from a successfully-verified piece of drand randomness.
+    /// This object can be converted into a uniform random integer.
+    struct Randomness has drop {
+        bytes: vector<u8>
+    }
+
+    /// Checks if the randomness in `bytes` verifies for the specified `round`.
+    /// If it verifies, returns the actual randomness, which is a hash function applied over `bytes`.
+    public fun verify_and_extract_randomness(
+        bytes: vector<u8>,
+        round: u64): Option<Randomness>
+    {
+        let pk = extract(&mut deserialize<G2, FormatG2Compr>(&DRAND_PUBKEY));
+        let sig = extract(&mut deserialize<G1, FormatG1Compr>(&bytes));
+        let msg_hash = hash_to<G1, HashG1XmdSha256SswuRo>(&DRAND_DST, &round_number_to_bytes(round));
+        assert!(eq(&pairing<G1, G2, Gt>(&msg_hash, &pk), &pairing<G1, G2, Gt>(&sig, &one<G2>())), 1);
+        option::some(Randomness {
+            bytes: sha3_256(bytes)
+        })
+    }
+
+    /// Returns a number in $[0, max)$ given some drand (verified) `randomness` and a `Type`.
+    /// Note: This is a one-shot API that consumes the `randomness`.
+    public fun random_number(randomness: Randomness, max: u64): u64 {
+        assert!(vector::length(&randomness.bytes) >= 8, error::invalid_argument(E_INCORRECT_RANDOMNESS));
+
+        let entropy = sha3_256(randomness.bytes);
+
+        // We can convert the 256 uniform bits in `randomness` into a uniform 64-bit number `w \in [0, max)` by
+        // taking the last 128 bits in `randomness` modulo `max`.
+        let last_128_bits = vector::trim(&mut entropy, 16);
+        let num : u128 = 0;
+        let max_128 = (max as u128);
+
+        // Ugh, we have to manually deserialize this into a u64 (8 chunks of 8 bits each)
+        while (!vector::is_empty(&last_128_bits)) {
+            let byte = vector::pop_back(&mut last_128_bits);
+            num = num << 8;
+            num = num + (byte as u128);
+        };
+
+        ((num % max_128) as u64)
```

Comment:
technically not uniform

Reply:
Yes, indeed. In fact, I should keep`num` 256 bits rather than 128 in order to minimize the bias...

Reply:
AFAICT, the bias is computationally-indistiguinshable, but let me know if you disagree. @zjma and I analyzed this a few months ago.

Reply:
the relative error is bounded by 2^-64. I agree it's unlikely we even have security guarantees that the cryptofgraphic hash functions won't have non-uniformity errors of such size. What I mostly meant is that some comment about the error analysis and compared with reasonable bounds on errors that it's very acceptable